### PR TITLE
Boulabytes

### PR DESCRIPTION
**Pull request by boulabytes**
Add customisable minimum stars matches
Add remove dark function
Add ui preferences objects, load / save functions

### CODE REVIEW

**Comment by boulabytes** [comment]
So, a friend and I ran some tests to validate dark removal.
It works, but we have to use a master dark (a stack of at least 50 darks sub) to really improve the stacking result. If we used a unique dark sub, it was better but not enough.
So we can say that this option is very usefull if we take time to build a well done master darks library.

File: src/als/ui/prefs_ui.ui [hunk 1]
```diff
@@ -209,6 +209,118 @@
...
+         <size>
+          <width>160</width>
+          <height>20</height>
+         </size>
+        </property>
+       </spacer>
+      </item>
+     </layout>
+    </widget>
+   </item>
```
Comment (deufrai) [inline src/als/ui/prefs_ui.ui:323]: globally on the preferences dialog: 

- We miss an accelerator on minimum stars spinbox
- I suggest the following section order. At least, 'misc' should be last :
 pathes, preprocessing, alignment, file saving, web server, misc

File: src/als/config.py [hunk 2]
```diff
@@ -38,6 +38,9 @@
 _IMAGE_SAVE_FORMAT = "image_save_format"
 _FULL_SCREEN = "full_screen"
 _WWW_REFRESH_PERIOD = "web_refresh_period"
+_AL_MINIMUM_STARS = "alignment_minimum_stars"
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/config.py:41]: You're right ! I gonna do the changes.

File: src/als/ui/dialogs.py [hunk 3]
```diff
@@ -123,6 +133,18 @@ def browse_work(self):
...
+    @log
+    def browse_dark(self):
+        """Opens a folder dialog to choose dark file"""
+        dark_file_path = QFileDialog.getOpenFileName(self,
+                                                    _("Select dark file"),
+                                                    self._ui.ln_master_dark_path.text())
+        if dark_file_path[0]:
+            self._ui.ln_master_dark_path.setText(dark_file_path[0])
+
+        #self._show_missing_folders()
```
[the remark this answers is not in the document]
Comment (deufrai) [reply src/als/ui/dialogs.py:146]: ... and called from here as well :)

[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/ui/dialogs.py:146]: Nop, this comment is a mistake ;) . I was debugging and commented out this line which obviously was not involved in the bug. Then I forget to comment it out :p
I gonna correct it.

File: src/als/config.py [hunk 4]
```diff
@@ -242,6 +248,71 @@ def set_scan_folder_path(path):
...
+    _set(_AL_MINIMUM_STARS, str(star_number))
+
+
+def set_use_master_dark(use_dark: bool):
+    """
+    Set use dark indicator
+
+    :param full: Remove master dark from images ?
+    :type full: bool
+    """
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/config.py:277]: Yeah, I was to hurry, and well... copy/past caveat :D.
Gonna correct it.

[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/config.py:277]: Done.

File: src/als/config.py [hunk 5]
```diff
@@ -242,6 +248,71 @@ def set_scan_folder_path(path):
...
+    return int(_get(_AL_MINIMUM_STARS))
+
+
+def set_al_minimum_stars(star_number):
+    """
+    Sets the alignment minimum stars value.
+
+    :param path: the minimum stars number for alignment
+    :type path: int
+    """
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/config.py:267]: Done.

File: src/als/processing.py [hunk 6]
```diff
@@ -376,6 +379,21 @@ def process_image(self, image: Image):
...
+class RemoveDark(ImageProcessor):
+    """
+    Provides image dark removal.
+    """
+
+    @log
+    def process_image(self, image: Image):
+
+        if config.get_use_master_dark():
+            masterdark = input.read_disk_image(Path(config.get_master_dark_file_path()))
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/processing.py]: Ok, I tried to do so. :)

File: src/als/processing.py [hunk 7]
```diff
@@ -376,6 +379,21 @@ def process_image(self, image: Image):
...
+            image.data = image.data - masterdark.data
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/processing.py]: I tried to correct it.

File: src/als/stack.py [hunk 8]
```diff
@@ -28,10 +28,10 @@
 from als.model.base import Image
 from als.model.data import STACKING_MODE_SUM, STACKING_MODE_MEAN
 from als.processing import QueueConsumer
-
+from als import config
 _LOGGER = logging.getLogger(__name__)
 
-_MINIMUM_MATCHES_FOR_VALID_TRANSFORM = 25
+_MINIMUM_MATCHES_FOR_VALID_TRANSFORM = config.get_al_minimum_stars()
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/stack.py]: Ok, I move it to a more convenient place.

File: src/als/ui/dialogs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/ui/dialogs.py]: Corrected.

File: src/als/ui/prefs_ui.ui [hunk 9]
```diff
@@ -209,6 +209,118 @@
...
+          </property>
+          <property name="buddy">
+           <cstring>spn_al_minimum_stars</cstring>
+          </property>
+         </widget>
+        </item>
+        <item row="0" column="1">
+         <widget class="QSpinBox" name="spn_al_minimum_stars">
+          <property name="minimum">
+           <number>1</number>
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/ui/prefs_ui.ui]: Change it to 5 which seems to work with my images.

File: src/als/ui/prefs_ui.ui (hunk 1 above)
Comment (boulabytes) [reply src/als/ui/prefs_ui.ui:307]: Ok, I sorted sections.

File: src/als/config.py [hunk 10]
```diff
@@ -242,6 +248,71 @@ def set_scan_folder_path(path):
...
+
+
+def get_master_dark_file_path():
+    """
+    Retrieves the master dark file path.
+
+    :return: the master dark file path
+    :rtype: str
+    """
+    return str(_get(_MASTER_DARK_FILE_PATH))
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/config.py]: done.

File: src/als/config.py [hunk 11]
```diff
@@ -242,6 +248,71 @@ def set_scan_folder_path(path):
...
+
+
+def set_master_dark_file_path(path):
+    """
+    Sets the master dark file path.
+
+    :param path: the master dark file path
+    :type path: str
+    """
+    _set(_MASTER_DARK_FILE_PATH, str(path))
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/config.py]: done.

File: src/als/ui/prefs_ui.ui (hunk 1 above)
Comment (boulabytes) [reply src/als/ui/prefs_ui.ui:307]: Hi, It seems I don't understand correctly ; could you explain what is an accelerator ?

Comment (deufrai) [reply src/als/ui/prefs_ui.ui:294]: Oh sorry : it is a way of quickly accessing a GUI item via keyboard (Alt+<key>)
Like what you did with the dark section, for example : Alt+U toggles the checkbox

**Comment by gehelem** [comment]
Just tested quiclky with my stock images
It looks to work quite well !
I let @deufrai going on with coding stuff...
Anyway, thank you for this work

File: src/als/ui/prefs_ui.ui (hunk 1 above)
Comment (boulabytes) [reply src/als/ui/prefs_ui.ui:294]: Ok, corrections done.

File: src/als/processing.py [hunk 12]
```diff
@@ -376,6 +379,35 @@ def process_image(self, image: Image):
...
+    Provides image dark removal.
+    """
+
+    @log
+    def process_image(self, image: Image):
+
+        if config.get_use_master_dark():
+            try:
+                masterdark = input.read_disk_image(Path(config.get_master_dark_file_path()))
+            except:
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/processing.py]: Ok, I add a test to verify data type and I modify the other tests to comply to your comments.

File: src/als/stack.py [hunk 13]
```diff
@@ -310,6 +310,7 @@ def _find_transformation(self, image: Image):
         :return: the found transformation
         :raises: StackingError when no transformation is found using the whole image
         """
+        _MINIMUM_MATCHES_FOR_VALID_TRANSFORM = config.get_align_minimum_stars()
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/stack.py]: Done.

File: src/als/ui/dialogs.py [hunk 14]
```diff
@@ -123,6 +133,18 @@ def browse_work(self):
...
+    @log
+    def browse_dark(self):
+        """Opens a folder dialog to choose dark file"""
+        dark_file_path = QFileDialog.getOpenFileName(self,
+                                                    _("Select dark file"),
+                                                    self._ui.ln_master_dark_path.text())
+        if dark_file_path[0]:
+            self._ui.ln_master_dark_path.setText(dark_file_path[0])
+
+        self._show_missing_folders()
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/ui/dialogs.py]: I modify the method, to check if the path actually exists.

File: src/als/ui/dialogs.py [hunk 15]
```diff
@@ -45,17 +50,17 @@ def __init__(self, parent=None):
...
+    def _validate_all_pathes(self):
         """
         Draw a red border around text fields containing a path to a missing folder
         """
 
-        for ui_field in [self._ui.ln_work_folder_path, self._ui.ln_scan_folder_path]:
+        for ui_field in [self._ui.ln_work_folder_path, self._ui.ln_scan_folder_path, self._ui.ln_master_dark_path]:
 
-            if not Path(ui_field.text()).is_dir():
+            if not Path(ui_field.text()).exists():
```
Comment (deufrai) [inline src/als/ui/dialogs.py:63]: Sorry if I sound like a PITA, but...
I think downgrading the 2 folder tests to only existence test is a loss.

We should actually check if the 2 expected folders are folders and if the expected file is a file

File: src/als/ui/dialogs.py [hunk 16]
```diff
@@ -121,7 +131,19 @@ def browse_work(self):
...
+    @log
+    def browse_dark(self):
+        """Opens a folder dialog to choose dark file"""
+        dark_file_path = QFileDialog.getOpenFileName(self,
+                                                     _("Select dark file"),
+                                                     self._ui.ln_master_dark_path.text())
+        if dark_file_path[0]:
+            self._ui.ln_master_dark_path.setText(dark_file_path[0])
+
+        self._validate_all_pathes()
```
Comment (deufrai) [inline src/als/ui/dialogs.py:146]: This call to self._validate_all_pathes() on line 146 is useless. Sorry, I misguided you on this one

File: src/als/ui/prefs_ui.ui [hunk 17]
```diff
@@ -91,12 +91,111 @@
...
+          <widget class="QCheckBox" name="chk_use_dark">
+           <property name="text">
+            <string>&amp;Use master dark</string>
+           </property>
+          </widget>
+        </item>
+        <item row="1" column="0">
+         <widget class="QLabel" name="lbl_dark_folder">
+          <property name="text">
+           <string>Master &amp;dark folder :</string>
```
Comment (deufrai) [inline src/als/ui/prefs_ui.ui:112]: this 'd' accelerator is already used (By 'Debug logs')

Comment (boulabytes) [reply src/als/ui/prefs_ui.ui:112]: Damned, didn't see it :P

File: src/als/ui/dialogs.py (hunk 15 above)
Comment (boulabytes) [reply src/als/ui/dialogs.py:63]: No, you're right.
I just thought that the only way, for example, to put a file in the scan folder was to edit manually the config file. But in fact changes may come from filesystem changes (i.e : the user transforms a folder into a file).
So yeah, I gonna change it.

Comment (deufrai) [reply src/als/ui/dialogs.py:63]: good spirit. We want to be as foolproof as possible

File: src/als/ui/prefs_ui.ui (hunk 17 above)
Comment (boulabytes) [reply src/als/ui/prefs_ui.ui]: Done

File: src/als/ui/dialogs.py (hunk 16 above)
Comment (boulabytes) [reply src/als/ui/dialogs.py]: Done ;)

File: src/als/ui/dialogs.py (hunk 15 above)
Comment (boulabytes) [reply src/als/ui/dialogs.py]: Done

File: src/als/ui/dialogs.py [hunk 18]
```diff
@@ -60,6 +65,11 @@ def _show_missing_folders(self):
             else:
                 ui_field.setStyleSheet("border: 1px")
 
+        if not Path(self._ui.ln_master_dark_path.text()).is_file():
+            self._ui.ln_master_dark_path.setStyleSheet("border: 1px solid orange")
+        else:
+            self._ui.ln_master_dark_path.setStyleSheet("border: 1px")
+
```
Comment (deufrai) [inline src/als/ui/dialogs.py:72]: Please don't duplicate stylesheet strings.
If one day we decide -- out of nowhere -- to change the warning color, we'll have to do it twice (or even more).

Comment (deufrai) [reply src/als/ui/dialogs.py:72]: And remember : an empy value must be allowed for the dark path

File: src/als/ui/prefs_ui.ui [hunk 19]
```diff
@@ -91,12 +91,111 @@
...
+     <property name="title">
+      <string>Alignment</string>
+     </property>
+     <layout class="QHBoxLayout" name="horizontalLayout_4">
+      <item>
+       <layout class="QGridLayout" name="gridLayout">
+        <item row="0" column="0">
+         <widget class="QLabel" name="lbl_minimum_match_count">
+          <property name="text">
+           <string>Minimum stars matches &amp;count :</string>
```
Comment (deufrai) [inline src/als/ui/prefs_ui.ui:155]: Please remove 'stars' word

File: src/als/ui/dialogs.py (hunk 18 above)
Comment (boulabytes) [reply src/als/ui/dialogs.py:75]: Ok, now it tests if chk_use_dark is checked and then it verifies if the path is correct. Dark are used only if chk_use_dark is checked so the verification is need there.

File: src/als/ui/prefs_ui.ui (hunk 19 above)
Comment (boulabytes) [reply src/als/ui/prefs_ui.ui]: Done.

File: src/als/stack.py [hunk 20]
```diff
@@ -335,10 +335,10 @@ def _find_transformation(self, image: Image):
                 matches_count = len(matches[0])
                 _LOGGER.debug(f"image matched features count : {matches_count}")
 
-                if matches_count < _MINIMUM_MATCHES_FOR_VALID_TRANSFORM:
+                if matches_count < minimum_matches_for_valid_transform:
                     _LOGGER.warning(f"Found transformation but matches count is too low : "
-                                    f"{matches_count} < {_MINIMUM_MATCHES_FOR_VALID_TRANSFORM}. "
-                                    "Discarding transformation")
+                                  f"{matches_count} < {minimum_matches_for_valid_transform}. "
```
Comment (deufrai) [inline src/als/stack.py:340]: please fix identation :)

We try and comply to PEP8. And for this specific case, please see https://stackoverflow.com/questions/15435811/what-is-pep8s-e128-continuation-line-under-indented-for-visual-indent

File: src/als/ui/dialogs.py (hunk 18 above)
Comment (deufrai) [reply src/als/ui/dialogs.py:75]: Imagine the following scenario 

- Last time ALS was used, dark file path was correct, but user chosed not to do dark subtraction and unchecked the box
- User exited ALS and did some other things, you know, for life :)
- Configured dark file got deleted from the user computer. These things do happen, you know
- User launches ALS & opens preferences
- As the checkbox is unchecked, the dark file path is not validated and no warning is shown
- User now decides to do dark subtraction and checks the box. She is now confident that no warning being shown, dark subtraction will occur correctly on next session
- User scratches her head as she keeps seeing those pesky colorful trails and eventually realizes something is fishy with the dark file when she hopefully-soon-enough opens preferences again.

I do think we should check the validity of the dark file path whenever said path is non-empty, regardless of the checkbox state.

Or do you have a better idea ?

File: src/als/ui/dialogs.py [hunk 21]
```diff
@@ -45,20 +50,28 @@ def __init__(self, parent=None):
...
-        self._show_missing_folders()
+        self._validate_all_pathes()
 
     @log
-    def _show_missing_folders(self):
+    def _validate_all_pathes(self):
         """
         Draw a red border around text fields containing a path to a missing folder
         """
+        warning_style = "border: 1px solid orange"
```
Comment (deufrai) [inline src/als/ui/dialogs.py:60]: Those are constants, please consider using CAPS and '_' prefix

Example : _MY_NICE_CONSTANT

File: src/als/stack.py (hunk 20 above)
Comment (boulabytes) [reply src/als/stack.py:340]: Fixed

File: src/als/ui/dialogs.py (hunk 18 above)
Comment (boulabytes) [reply src/als/ui/dialogs.py:75]: No, you're right :) . 
So I modify the code. Now it verifies path validity systematically if a path is set. Empty path is allowed only if chk_use_dark is not checked.

File: src/als/ui/dialogs.py (hunk 21 above)
Comment (boulabytes) [reply src/als/ui/dialogs.py]: Done.

File: src/als/ui/dialogs.py [hunk 22]
```diff
@@ -45,20 +51,27 @@ def __init__(self, parent=None):
...
             else:
-                ui_field.setStyleSheet("border: 1px")
+                ui_field.setStyleSheet(_NORMAL_STYLE_SHEET)
+
+        if self._ui.chk_use_dark.isChecked():
+            if not Path(self._ui.ln_master_dark_path.text()).is_file():
+                self._ui.ln_master_dark_path.setStyleSheet(_WARNING_STYLE_SHEET)
+        else:
+            if (not Path(self._ui.ln_master_dark_path.text()).is_file()) and (self._ui.ln_master_dark_path.text() != ""):
+                self._ui.ln_master_dark_path.setStyleSheet(_WARNING_STYLE_SHEET)
```
[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/ui/dialogs.py]: Again you're right, it seems my coding skill is very rusty :) . But, it's not my actual job ;) .
I gonna correct it :) .

[the remark this answers is not in the document]
Comment (boulabytes) [reply src/als/ui/dialogs.py]: Alright, I tried to implent it in the right way and add the qt auto-connect method you gave me.

[the remark this answers is not in the document]
Comment (deufrai) [reply src/als/ui/dialogs.py]: > Again you're right, it seems my coding skill is very rusty :) . But, it's not my actual job ;) .
> I gonna correct it :) .

Don't worry. You show other skills here : perseverance and will to improve

**Comment by deufrai** [comment]
After lots of fruitful discussions, this will be merged in the morning
Good work & Thanks again !!!

**Pull request merged**